### PR TITLE
Mount the SSH keys as secrets in all workspace containers

### PR DESCRIPTION
**Pull request by vparfonov**
Signed-off-by: Vitalii Parfonov <vparfono@redhat.com>

### What does this PR do?
Mount the SSH keys as secrets in all workspace containers.
How it works:

- wsmaster is responsible for putting the user ssh keys to the k8s [secret](https://kubernetes.io/docs/concepts/configuration/secret);
- kubernetes is responsible for mount the k8s secret as file to each workspace containers
 
Then starting a workspace, we will grep all SSH Keys registered for VCS and create K8s Secret for
them. All created secrets will be mounted on each container by path  `/etc/ssh/{sshKeyName}/ssh-privatekey` as a file. ([Secret Type for SSH keys](https://github.com/kubernetes/kubernetes/blob/7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/pkg/apis/core/types.go#L4458)).
Also will be created and mounted ConfigMap with SSH settings to the `/etc/ssh/ssh_config`.

 

### What issues does this PR fix or reference?
#13520

### CODE REVIEW

**Comment by musienko-maxim** [comment]
Can one of the admins verify this PR?

**Comment by rhopp** [comment]
ci-test

**Review by l0rd — APPROVED** [review]
Big +1 for including this in 7.0.0 (N.B. I haven't reviewed the content but just the description)

**Review by skabashnyuk — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+                  .withSubPath(SSH_CONFIG)
+                  .withReadOnly(false)
+                  .withNewReadOnly(false)
+                  .build();
+          container.getVolumeMounts().add(volumeMount);
+        });
+  }
+
+  /**
+   * @param host the host of the host
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: can you rephrase this?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                      Base64.getEncoder().encodeToString(sshPair.getPrivateKey().getBytes()))
+                  .withType(SECRET_TYPE_SSH)
+                  .withNewMetadata()
+                  .withName(getValidNameForSecret(sshPair.getName()))
+                  .endMetadata()
+                  .build();
+
+          k8sEnv.getSecrets().put(secret.getMetadata().getName(), secret);
+
+          if (!sshPairs.isEmpty()) {
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: This check seems redundant for me, isn't we already checked this in line 96?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                .withName("sshconfigmap")
+                .endMetadata()
+                .withData(map)
+                .build();
+
+        k8sEnv.getConfigMaps().put(configMap.getMetadata().getName(), configMap);
+        k8sEnv.getPodsData().values().forEach(p -> addConfigFile(p.getSpec()));
+      }
+    } catch (ServerException e) {
+      LOG.warn("Unable get SSH Keys ", e);
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: Maybe it makes sense to explain message a bit

**Review by mshaposhnik — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 2 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: yes. thanks

**Review by rhopp — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+              .forEach(p -> addSshKeySecret(secret.getMetadata().getName(), p.getSpec()));
+
+          sshConfig.append(buildConfig(sshPair.getName()));
+        }
+
+        map.put(SSH_CONFIG, sshConfig.toString());
+        ConfigMap configMap =
+            new ConfigMapBuilder()
+                .withNewMetadata()
+                .withName("sshconfigmap")
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: The same value is reused inside of `addConfigFile`. Please consider using variable or constant for it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: +1. Something like 
```
 LOG.warn("Unable get SSH Keys. Cause: %s", e.getMessage(), e);
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+          container.getVolumeMounts().add(volumeMount);
+        });
+  }
+
+  private void addConfigFile(PodSpec podSpec) {
+    podSpec
+        .getVolumes()
+        .add(
+            new VolumeBuilder()
+                .withName("configvolume")
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: I think `configvolume` is too abstract name for volume that may cause a collision with user-defined volumes.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                  .withSubPath(SSH_CONFIG)
+                  .withReadOnly(false)
+                  .withNewReadOnly(false)
+                  .build();
+          container.getVolumeMounts().add(volumeMount);
+        });
+  }
+
+  /**
+   * @param host the host of version control service (e.g. github.com, gitlab.com and etc)
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java:214]: Maybe the following format would be better 
```java
  /**
   * Returns the ssh configuration entry which includes host and identity file location
   * <p>Example of provided configuration:
   * <pre>
   * host github.com
   * HostName github.com
   * IdentityFile /etc/ssh/github-com/ssh-privatekey
   * </pre>
   *
   * @param host the host of version control service (e.g. github.com, gitlab.com and etc)
   * @return the ssh configuration which include host and identity file location
   */
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+   * @return the ssh configuration which include host of host and identity file location
+   *     <p>Example of provided configuration:
+   *     <p>host github.com HostName github.com IdentityFile /etc/ssh/github-com/ssh-privatekey
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: Is it true? I see `\n` in code, it's only one line example.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        .append("\n")
+        .append("HostName ")
+        .append(host)
+        .append("\nIdentityFile ")
+        .append(SSH_BASE_CONFIG_PATH)
+        .append(validName)
+        .append("/")
+        .append(SSH_PRIVATE_KEY)
+        .append("\n");
+    return stringBuilder.toString();
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: IDE propose me to replace it with String, like
```java
    return "host "
        + host
        + "\n"
        + "HostName "
        + host
        + "\nIdentityFile "
        + SSH_BASE_CONFIG_PATH
        + getValidNameForSecret(host)
        + "/"
        + SSH_PRIVATE_KEY
        + "\n";
```
WDYT?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  /**
+   * Check is given name available for secret name
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: I think java doc is a bit misleading. In my understanding `Check` should return true/false or throw an exception in case of invalid value. But this method transforms any value to a valid one.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+   *
+   * @param name
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: Please remove empty items @param and @return or write there something

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    Map<String, String> map = new HashMap<>();
+
+    try {
+      List<SshPairImpl> sshPairs = sshManager.getPairs(identity.getOwnerId(), "vcs");
+      if (!sshPairs.isEmpty()) {
+        for (SshPair sshPair : sshPairs) {
+          if (isNullOrEmpty(sshPair.getName()) || isNullOrEmpty(sshPair.getPrivateKey())) {
+            continue;
+          }
+          Secret secret =
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: Quite much low-level and high-level logic code are mixed here. I propose to move low-level code to private methods, like
```java
....
 public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
      throws InfrastructureException {
    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);

    try {
      List<SshPairImpl> sshPairs = sshManager.getPairs(identity.getOwnerId(), "vcs");
      if (sshPairs.isEmpty()) {
        return;
      }

      StringBuilder sshConfig = new StringBuilder();

      for (SshPair sshPair : sshPairs) {
        doProvisionSshKey(sshPair, k8sEnv);

        sshConfig.append(buildConfig(sshPair.getName()));
      }

      doProvisionSshConfig(sshConfig.toString(), k8sEnv);
    } catch (ServerException e) {
      LOG.warn("Unable get SSH Keys. Cause: %s", e.getMessage(), e);
    }
  }

  private void doProvisionSshKey(SshPair sshPair, KubernetesEnvironment k8sEnv) {
  ...
  }

  private void doProvisionSshConfig(String sshConfig, KubernetesEnvironment k8sEnv) {
  ...
  }
```

**Review by sleshchenko — COMMENTED** [review]
I believe I have fixes for many of my comments, you can reuse it https://gist.github.com/sleshchenko/15f31113206f270d839172c335046da5

In general looks good, let me know when all comments will be addressed and I'll do another round of review.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 8 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: +1

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 10 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: yep, you are right

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 11 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: thanks

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+import org.eclipse.che.commons.annotation.Traced;
+import org.eclipse.che.commons.tracing.TracingTags;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class allows mount user SSH Keys as <a
+ * href="https://kubernetes.io/docs/concepts/configuration/secret">Kubernetes Secrets</a> on each
+ * workspace containers. <br>
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+ * <strong>How it works</strong> <br>
+ * When starting a workspace, we will grep all SSH Keys registered for VCS and create K8s Secret the
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+  @Override
+  @Traced
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);
+
+    StringBuilder sshConfig = new StringBuilder();
+    Map<String, String> map = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 9 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 3 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: done

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 7 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: Looks like `containers` are still there

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+import org.eclipse.che.commons.tracing.TracingTags;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class allows mount user SSH Keys as <a
+ * href="https://kubernetes.io/docs/concepts/configuration/secret">Kubernetes Secrets</a> on each
+ * workspace containers. <br>
+ * <strong>How it works</strong> <br>
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: I propose rewrite `How it works` section in the following way:
```java
 * <p><strong>How it works:</strong>
 * <ul>
 * <li>all SSH Keys registered for VCS are fetched;</li>
 * <li>create K8s Secrets with <a href=https://github.com/kubernetes/kubernetes/blob/7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/pkg/apis/core/types.go#L4458">SSH Key type</a> 
 * for each of SSH keys;</li>
 * <li>secrets are mounted on each container by path '/etc/ssh/{sshKeyName}/ssh-privatekey' as a file;</li>
 * <li>ConfigMap with SSH settings is created and mounted to container by path '/etc/ssh/ssh_config'.</li>
 * </ul>
```
WDYT?

**Review by sleshchenko — APPROVED** [review]
LGTM

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java (hunk 15 above)
Comment (vparfonov) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java]: agree looks better, thx
